build/net/src/lib.rs

```diff
@@ -82,23 +98,129 @@ pub fn load_net_config() -> Result<NetConfig> {
     Ok(cfg)
 }
 
-pub fn generate_vlan_consts(
+pub fn generate_port_consts(
     config: &NetConfig,
     mut out: impl std::io::Write,
 ) -> Result<(), std::io::Error> {
-    let vlan = config.vlan.unwrap();
-    let end = vlan.start + vlan.count;
-    if end > 0xFFF {
-        panic!("Invalid VLAN range (must be < 4096)");
+    let ports = config
+        .vlans
+        .values()
+        .map(|v| v.port)
+        .collect::<BTreeSet<_>>()
+        .len();
+    assert!(ports <= 2);
+    let port_count = ports.max(1);
+    let s = quote! {
+        pub const PORT_COUNT: usize = #port_count;
+    };
+    writeln!(out, "{s}")
+}
+
+fn check_vlan_config(config: &NetConfig) {
+    if config.vlans.values().any(|v| v.vid > 0xFFF) {
+        panic!("Invalid VLAN VID (must be < 4096)");
     }
-    writeln!(
-        out,
-        "
-pub const VLAN_RANGE: core::ops::Range<u16> = {:#x}..{:#x};
-pub const VLAN_COUNT: usize = {};
-",
-        vlan.start, end, vlan.count
-    )
+    for (k1, v1) in &config.vlans {
+        for (k2, v2) in &config.vlans {
+            if k1 != k2 && v1.vid == v2.vid {
+                panic!("Duplicate VID {} (in {k1} and {k2})", v1.vid);
+            }
+        }
+    }
+}
+
+pub fn generate_vlan_consts(
+    config: &NetConfig,
+    mut out: impl std::io::Write,
+) -> Result<(), std::io::Error> {
+    check_vlan_config(config);
+
+    let vlan_count = config.vlans.len();
+    let vids = config.vlans.values().map(|cfg| cfg.vid).collect::<Vec<_>>();
+    let s = quote! {
+        #[allow(unused)]
+        pub const VLAN_COUNT: usize = #vlan_count;
+
+        #[allow(unused)]
+        pub const VLAN_VIDS: [u16; #vlan_count] = [
+            #(
+                #vids
+            ),*
+        ];
+    };
+    writeln!(out, "{s}")
+}
+
+pub fn generate_vlan_enum(
+    config: &NetConfig,
+    mut out: impl std::io::Write,
+) -> Result<(), std::io::Error> {
+    check_vlan_config(config);
+
+    let s = if config.vlans.is_empty() {
+        quote! {
+            #[derive(
+                Copy, Clone, Eq, PartialEq,
+                enum_map::Enum,
+                serde::Serialize, serde::Deserialize,
+                hubpack::SerializedSize,
+            )]
```

Comment:
nitpicky, feel free to ignore me: maybe the code inside the `if` could *just* make the list of varriants, and we could `quote!` that list of variants outside of the if with the rest of the enum definition, like:
```rust
let vlans = if config.vlans.is_empty() {
    quote! { None, }
} else {
   // ... the rest of the stuff ...
};
let s = quote! {
    #[derive(
        Copy, Clone, Eq, PartialEq,
        enum_map::Enum,
        serde::Serialize, serde::Deserialize,
        hubpack::SerializedSize,
    )]
    pub enum VLanId {
        #(vlans)
    }
};
```
that way, we don't have the opportunity for annoying bugs where we add an additional derive, and forget to add it to the "no VLANs" case, or something.

this is probably not super important, but it's the kind of thing i think too hard about...

Reply:
I'm going to pass on this because the two branches don't just define the `enum VLanId`; the non-empty branch also adds an impl block, so splitting them is trickier.